### PR TITLE
net: sockets: tls: Fix TLS/DTLS client POLLHUP notification

### LINKED ISSUE #37720: net: dtls: received close_notify alerts are not properly handled by DTLS client sockets
**Describe the bug**
In the current state of the DTLS implementation, received `close_notify` alerts are not properly handled by a DTLS client socket.

According to the mbedtls documentation the SSL context must be reset when `mbedtls_ssl_read ` returns the following:

> If this function returns something other than a positive value, MBEDTLS_ERR_SSL_WANT_READ, MBEDTLS_ERR_SSL_WANT_WRITE, MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS, MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS or MBEDTLS_ERR_SSL_CLIENT_RECONNECT, you must stop using the SSL context for reading or writing, and either free it or call mbedtls_ssl_session_reset() on it before re-using it for a new connection; the current connection must be closed.
see https://tls.mbed.org/api/ssl_8h.html#aa2c29eeb1deaf5ad9f01a7515006ede5

For the case of `MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY` the SSL context is not reset for DTLS clients: https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v2.6.0/subsys/net/lib/sockets/sockets_tls.c#L1808
This may lead to an application still trying to send data over an already closed connection.

Furthermore, it is not possible to detect (in an application) that the session has been closed when using `poll()` because neither `POLLHUP` nor `POLLERR` are set.

**To Reproduce**
1. open a DTLS client socket in zephyr
2. establish session with a DTLS server
3. send `close_notify` alert from DTLS server

**Expected behavior**

- Internally in `sockets_tls`, the SSL context should be reset when receiving a `close_notify` alert on a DTLS client socket (similar to how it is already done for DTLS server sockets: https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v2.6.0/subsys/net/lib/sockets/sockets_tls.c#L1893).

- Ideally, `POLLHUP` is set as well on the poll FD so an application can decide how to deal with the closed session.

**Impact**
Possibly a lot of packets are sent over a closed connection unnecessarily until e.g. an application decides to renew the DTLS session.

**Environment (please complete the following information):**
 - Zephyr v2.6.0

### PR DESCRIPTION
**Pull request by rlubos**
The `poll()` function did not report POLLHUP if the peer ended the DTLS
session, making it impossible to detect such event on the application
side.

On the other hand, TLS erroneously reported POLLHUP along with each
POLLIN event, as the 0 returned by the `recv()` socket call was
wrongly interpreted (it was expected to get 0 in return as 0 bytes were
requested).

Fix this by introducing a helper function to process the mbedtls context
and verify if new application data is pending. Use this new function in
the poll handler, instead of a socket `recv()` call, to remove any
ambiguity in the usage, for both TLS and DTLS.

Fixes #37720

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

### CODE REVIEW

**Comment by chrta** [comment]
@rlubos Thank you for the fast response. @phise is out of office this week. I briefly tested this in our environment (nrf91, nrf-sdk 1.6.0/zephyr 2.6.0) with dtls using psk and it seems to behave fine now. The application is able to catch an error on the dtls socket and can reconnect, e.g. if the dtls server requests a renegotiation, which is not supported be mbedtls or zephyr. In this case the application can now detects the issue, close the socket and reopen it.

I do not feel comfortable enough with dtls and mbedtls to provide a code review, but as far as i can tell it looks good.

**Review by jukkar — COMMENTED** [review]
The changes look ok. The indentation of the code looks weird, could be the browser showing things incorrectly but could you check that.

**Comment by phise** [comment]
@rlubos Thanks for the fix. Tested it with a DTLS client application on the qemu_x86 target (and OpenSSL as a DTLS server). 

On reception of a close_notify alert:
- the internal session context is properly reset, a new handshake is initiated on the next sendto() call
- the application now recognizes that POLLHUP has been set on the corresponding FD

The changes look good to me, but I'm also no expert regarding DTLS.

**Comment by rlubos** [comment]
> The changes look ok. The indentation of the code looks weird, could be the browser showing things incorrectly but could you check that.

@jukkar Hmm, I don't see any issues on my side, could you post some example where it seems broken so I could double check that particular piece of code?

**Comment by jukkar** [comment]
> Hmm, I don't see any issues on my side, could you post some example where it seems broken so I could double check that particular piece of code?

I am just not seeing tabs being indented properly in my browser (firefox) when reviewing the code. Not sure what is going here and I am seeing the same issue in other PRs so just ignore my indentation comment for now.

**Comment by pfalcon** [comment]
> I am just not seeing tabs being indented properly in my browser (firefox) when reviewing the code. Not sure what is going here and I am seeing the same issue in other PRs so just ignore my indentation comment for now.

Ack. It looks as if recent Firefox started to use tabs of size 4 instead of 8. I'm using Firefox 91.0.2, and previously there was no such issue. Of course, the issue could start to happen on Github side too. (There's no such issue in Chromium 92.0.)

**Review by pfalcon — COMMENTED** [review]
I'd like to remind the concerns raised when implementing "TLS sockets" - TLS is effectively an application-level protocol, so shoveling it under transport layer of POSIX BSD Sockets is a layering violation (and thus a mistake). Various issues with that trickle since that. This particular patch seems to try to redefine meaning of POSIX socket error codes to deal with TLS "sockets" peculiarities. I'd recommend to reconsider this approach, don't use adhoc redefinition of error codes, and implement the needed functionality in other way (e.g. store and check more internal state to emulate proper socket semantics).

File: subsys/net/lib/sockets/sockets_tls.c
Comment (chrta) [reply subsys/net/lib/sockets/sockets_tls.c:882]: >  Assuming this translates to the actual socket errno

It does

>  there's no listening socket at the corresponding destination address/port. Existing ECONNABORTED seems like much better match for "TLS handshake error".

This errno (from this line) is also used when the client tries to initiate a dtls session and no server is running at all. At least in that case ECONNREFUSED seems to be better that ECONNABORTED.

Comment (pfalcon) [reply subsys/net/lib/sockets/sockets_tls.c:882]: > This errno (from this line) is also used when the client tries to initiate a dtls session and no server is running at all. At least in that case ECONNREFUSED seems to be better than ECONNABORTED.

Then it would be better to do more detailed translation of mbedTLS -> POSIX errno error codes, because otherwise changes like that improve one case but may lead to more confusion with other cases in the future (for example, if there's cipher/key mismatches and connection can't proceed due to that, then returning ECONNREFUSED for it is in no way better, because ECONNREFUSED means "no server running at all").

At the very least, that should go into a separate commit with the above explanation why it was done, so someone later could improve it (instead of risking that it may be just flip-flopped on seeing a case like described above).

File: subsys/net/lib/sockets/sockets_tls.c
Comment (rlubos) [reply subsys/net/lib/sockets/sockets_tls.c:1807]: I can't return 0 for datagram socket as for stream, because that would be ambigous (it's ok to get 0 on datagram recv). And I won't introduce a custom errno value to indicate that server closed the session, IMO it'd be more confusing that using something from the standard errno space (especially that it was recently unified accross standard libraries we support in Zephyr).

`ECONNABORTED` seemed like a good choice to me to indicate that peer closed the session, especially that I changed other uses of this value to avoid ambiguity (as explained in the commit message). Alternatively, an option would be to use `ENOTCONN` to indicate that socket is no longer connected.

File: subsys/net/lib/sockets/sockets_tls.c
Comment (rlubos) [reply subsys/net/lib/sockets/sockets_tls.c:1871]: The above function would fail if `mbedtls_ssl_session_reset()` fails, `ECONNABORTED` was used incorrectly as from what I've seen the function fails if it's not able to re-allocate resources from the heap, hence `ENOMEM`.

File: subsys/net/lib/sockets/sockets_tls.c
Comment (pfalcon) [reply subsys/net/lib/sockets/sockets_tls.c:882]: >>     because ECONNREFUSED means "no server running at all"

> Where does that come from? errno.h states:

That comes from deeply ingrained experience with standard sockets.

> #define ECONNREFUSED 111 /**< Connection refused */
> And a more detailed description from connect() man pages:
>       ECONNREFUSED
>              The target address was not listening for connections or refused the connection request.

To me, that translates to: "server was listening, but had its listen() backlog full".

> Yes, the handshake may fail for multiple reasons, but in my opinon "Connection refused" is a good summary that the connection did not succeed.

Again, with my socket experience, I'd find such error confusing (literally, would likely be misguided by it, thinking that there's a problem with server just accepting connections, not with TLS handshake). As I mentioned, one can only so much shovel TLS features into BSD Sockets semantics, so there always will be compromises. I tried to argue a point that this one isn't the best. I didn't try to -1 though, because in the end, only you, as the TLS sockets maintainer, can decide what makes sense with the overall picture in mind.

File: subsys/net/lib/sockets/sockets_tls.c
Comment (pfalcon) [reply subsys/net/lib/sockets/sockets_tls.c:1807]: > And I won't introduce a custom errno value to indicate that server closed the session, IMO it'd be more confusing that using something from the standard errno space

Yes (it would have been better to go with a separate API in that case).

> Alternatively, an option would be to use ENOTCONN to indicate that socket is no longer connected.

IMHO, that would be a better choice.

**Comment by rlubos** [comment]
I've pushed an updated version, ENOTCONN is returned when DTLS session is over and removed controversial `errno` change to get things done.

**Review by pfalcon — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Comment by chrta** [comment]
Is there something that blocks this merge? I would like to have this available in zephyr 2.7, if possible.

**Comment by pfalcon** [comment]
> Is there something that blocks this merge? I would like to have this available in zephyr 2.7, if possible.

Nothing AFAICT, it's just the pre-release freeze is already in effect, so merges are performed by the release manager/team based on the overall release preparation situation.

**Pull request merged**
